File: AddTwoNumbers.java
import structs.ListNode;

public class AddTwoNumbers {
  public static void main(String[] args) {
    AddTwoNumbers s = new AddTwoNumbers();
    ListNode l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
    ListNode l2 = new ListNode(5, new ListNode(6, new ListNode(4)));

    ListNode result = s.addTwoNumbers(l1, l2);
    while(result != null){
      System.out.print(result.val + "->");
      result = result.next;
    }
    System.out.print("\n");
  }

  public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    ListNode dummy = new ListNode(-1);
    ListNode copy = dummy;
    int carry = 0;
    while (l1 != null || l2 != null) {
      int sum = 0;
      if (l1 != null) {
        sum += l1.val;
        l1 = l1.next;
      }

      if (l2 != null) {
        sum += l2.val;
        l2 = l2.next;
      }

      sum += carry;
      copy.next = new ListNode(sum % 10);
      copy = copy.next;

      carry = sum / 10;
    }
    if (carry != 0) {
      copy.next = new ListNode(carry);
    }
    return dummy.next;
  }
}

File: add_two_numbers.go
package main

import "fmt"

type ListNode struct{
	Val int
	Next *ListNode
}

func main(){
	head_1 := &ListNode{2, &ListNode{4, &ListNode{3, nil}}}
	head_2 := &ListNode{5, &ListNode{6, &ListNode{4, nil}}}

	res := addTwoNumbers(head_1, head_2)
	for res != nil{
		fmt.Printf("%d\t", res.Val)
		res = res.Next
	}
	fmt.Print("\n")
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    var res_head *ListNode = &ListNode{-1, nil}
		var res_copy *ListNode = res_head
		var carry, sum int = 0, 0
		for l1 != nil || l2 != nil{
			sum = 0
			if(l1 != nil){
				sum = sum + l1.Val
				l1 = l1.Next
			}

			if(l2 != nil){
				sum = sum + l2.Val
				l2 = l2.Next
			}

			if carry != 0{
				sum = sum + carry
			}
			res_head.Next = &ListNode{sum%10, nil}
			res_head = res_head.Next
			carry = sum / 10
		}

		if carry != 0{
			res_head.Next = &ListNode{carry, nil}
		}

		return res_copy.Next
}